lightning/src/routing/network_graph.rs

```diff
@@ -0,0 +1,1672 @@
+		write!(f, "lowest_inbound_channel_fees: {:?}, channels: {:?}, announcement_info: {:?}",
+		   self.lowest_inbound_channel_fees, &self.channels[..], self.announcement_info)?;
+		Ok(())
+	}
+}
+
+impl Writeable for NodeInfo {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		(self.channels.len() as u64).write(writer)?;
+		for ref chan in self.channels.iter() {
+			chan.write(writer)?;
+		}
+		self.lowest_inbound_channel_fees.write(writer)?;
+		self.announcement_info.write(writer)?;
+		Ok(())
+	}
+}
+
+const MAX_ALLOC_SIZE: u64 = 64*1024;
+
+impl Readable for NodeInfo {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<NodeInfo, DecodeError> {
+		let channels_count: u64 = Readable::read(reader)?;
+		let mut channels = Vec::with_capacity(cmp::min(channels_count, MAX_ALLOC_SIZE / 8) as usize);
+		for _ in 0..channels_count {
+			channels.push(Readable::read(reader)?);
+		}
+		let lowest_inbound_channel_fees = Readable::read(reader)?;
+		let announcement_info = Readable::read(reader)?;
+		Ok(NodeInfo {
+			channels,
+			lowest_inbound_channel_fees,
+			announcement_info,
+		})
+	}
+}
+
+/// Represents the network as nodes and channels between them
+#[derive(PartialEq)]
+pub struct NetworkGraph {
+	channels: BTreeMap<u64, ChannelInfo>,
+	nodes: BTreeMap<PublicKey, NodeInfo>,
+}
+
+impl Writeable for NetworkGraph {
+	fn write<W: Writer>(&self, writer: &mut W) -> Result<(), ::std::io::Error> {
+		(self.channels.len() as u64).write(writer)?;
+		for (ref chan_id, ref chan_info) in self.channels.iter() {
+			(*chan_id).write(writer)?;
+			chan_info.write(writer)?;
+		}
+		(self.nodes.len() as u64).write(writer)?;
+		for (ref node_id, ref node_info) in self.nodes.iter() {
+			node_id.write(writer)?;
+			node_info.write(writer)?;
+		}
+		Ok(())
+	}
+}
+
+impl Readable for NetworkGraph {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<NetworkGraph, DecodeError> {
+		let channels_count: u64 = Readable::read(reader)?;
+		let mut channels = BTreeMap::new();
+		for _ in 0..channels_count {
+			let chan_id: u64 = Readable::read(reader)?;
+			let chan_info = Readable::read(reader)?;
+			channels.insert(chan_id, chan_info);
+		}
+		let nodes_count: u64 = Readable::read(reader)?;
+		let mut nodes = BTreeMap::new();
+		for _ in 0..nodes_count {
+			let node_id = Readable::read(reader)?;
+			let node_info = Readable::read(reader)?;
+			nodes.insert(node_id, node_info);
+		}
+		Ok(NetworkGraph {
+			channels,
+			nodes,
+		})
+	}
+}
+
+impl std::fmt::Display for NetworkGraph {
+	fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+		write!(f, "Network map\n[Channels]\n")?;
+		for (key, val) in self.channels.iter() {
+			write!(f, " {}: {}\n", key, val)?;
+		}
+		write!(f, "[Nodes]\n")?;
+		for (key, val) in self.nodes.iter() {
+			write!(f, " {}: {}\n", log_pubkey!(key), val)?;
+		}
+		Ok(())
+	}
+}
+
+impl NetworkGraph {
+	/// Returns all known valid channels
```

Comment:
Please specify the return types.

Reply:
I don't understand this one... Docs usually capture return values pretty [well](https://docs.rs/lightning/0.0.11/lightning/ln/router/struct.Router.html#method.get_addresses) automatically. What do you think is unclear here?

Reply:
What does the u64 in the hashmap represent, mostly.